Embed TouchActionsPanel directly into the Android editor UI

syntaxerror247 opened the pull request:
This PR embeds the `TouchActionsPanel` directly into the Android editor UI. Users can still switch to the `floating panel` by selecting the option in the editor setting `interface/touchscreen/touch_actions_panel`.


https://github.com/user-attachments/assets/a2847338-e6a1-4766-9a96-b26595e19a3c

syntaxerror247 committed:
Embed TouchActionsPanel directly into the Android editor UI

m4gr3d commented on editor/editor_settings.cpp:
```diff
@@ -584,17 +584,18 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
+	// Only available in the Android editor.
+	String touch_actions_panel_hints = "Disabled:0,Embedded Panel:1,Floating Panel:2";
+	EDITOR_SETTING_BASIC(Variant::INT, PROPERTY_HINT_ENUM, "interface/touchscreen/touch_actions_panel", 1, touch_actions_panel_hints)
+	set_restart_if_changed("interface/touchscreen/touch_actions_panel", true);
```
This could probably be done without restarting; for example you could listen to the settings changed event and free the panel if disabled, remove it from its parent if floating, or add it to its parent if embedded.

syntaxerror247 replied:
Yeah, that's definitely doable. It would require keeping a reference to the panel and the new main_hbox, then listening for setting changes to update the layout dynamically.
I figured users wouldn’t be changing that setting often, so I went with the simplest approach -- just requiring a restart.

That said, I can make the changes if that's prefered. Let me know what you think!

m4gr3d replied:
I expect users to play with it for a bit to figure how well it fits their workflow, but yes not touching afterward.

Despite this expected usage pattern, making it configurable without restarting is worth doing as it aligns with our current pattern for UI in the editor (e.g: on desktop users can float / dock panels without restarting the editor).

m4gr3d reviewed (commented):
Can you also provide the option to select whether the panel be on the left or right side of the editor window when embedded.

syntaxerror247 commented:
@m4gr3d I've updated the PR to include both of your suggestions. It no longer requires a restart to change the panel mode, and I've also added a button to switch the panel between the left and right sides.

m4gr3d commented on editor/editor_node.cpp:
```diff
@@ -7063,6 +7068,33 @@ void EditorNode::set_unfocused_low_processor_usage_mode_enabled(bool p_enabled)
+			break;
+		case 0:
+			if (touch_actions_panel != nullptr) {
+				touch_actions_panel->queue_free();
```
You should set `touch_actions_panel = nullptr` here.

m4gr3d reviewed (approved):
Looks great!

m4gr3d commented:
@syntaxerror247 Have you tried these changes on the XR editor on your Pico device? 
I'm wondering whether we should keep the feature enabled for the XR editor or not. 

cc @dsnopek @devloglogan

syntaxerror247 commented:
Yeah, I’ve tested it with the XR editor as well, and it works there without any issues.

I’d say to keep it enabled there too. When using with controller, the same limitations exist. Plus, connecting a keyboard hides the panel anyway, so I don’t see any downside to leaving it enabled in XR.

m4gr3d commented:
> Yeah, I’ve tested it with the XR editor as well, and it works there without any issues.
> 
> I’d say to keep it enabled there too. When using with controller, the same limitations exist. Plus, connecting a keyboard hides the panel anyway, so I don’t see any downside to leaving it enabled in XR.

Sounds good!

syntaxerror247 commented:
Updated the editor setting description to also mention XR editor.

llama-nl commented:
@syntaxerror247 could you look at this: https://github.com/godotengine/godot-proposals/issues/12220

Repiteo commented:
Thanks!
